### PR TITLE
Added checks to ensure new max and threshold supply values are greate…

### PR DESCRIPTION
**Pull request by DengreSarthak**
Added checks to ensure new max and threshold supply values are greater than the current total supply.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved validation to prevent setting the maximum or threshold token supply below the current total supply, ensuring better consistency and reliability when adjusting token limits.
  - Enhanced transfer restrictions to tighten conditions under which token transfers are blocked, improving security and compliance.
  - Updated error handling with clearer, custom error messages for better user feedback during token operations.
- **New Features**
  - Added explicit role management functions for granting and revoking minter roles, enhancing administrative control.
- **Tests**
  - Expanded test coverage for minting, burning, transfer restrictions, role management, and pagination features.
  - Added comprehensive edge case tests to ensure robust token behavior and access control enforcement.

### CODE REVIEW

File: contracts/ContributionAccountingToken.sol [hunk 1]
```diff
@@ -7,11 +7,20 @@ import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
...
     function onMinterRoleGranted(address minter) external;
 }
 
 contract ContributionAccountingToken is ERC20Burnable, ERC20Permit, AccessControl {
+    // Custom errors
+    error ExceedsMaxMintableAmount(uint256 requested, uint256 available);
+    error NewMaxSupplyNotLess(uint256 newMax, uint256 currentMax);
+    error NewMaxSupplyBelowTotal(uint256 newMax, uint256 total);
+    error NewMaxSupplyBelowThresholdSupply(uint256 newMax, uint256 thresholdSupply);
+    error NewThresholdNotLess(uint256 newThreshold, uint256 currentThreshold);
```
Comment (ceilican) [inline contracts/ContributionAccountingToken.sol:19]: `ThresholdNotDecreased` would be a clearer name for this error.

File: contracts/ContributionAccountingToken.sol [hunk 2]
```diff
@@ -7,11 +7,20 @@ import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
...
+    error NewThresholdBelowTotal(uint256 newThreshold, uint256 total);
+    error NewMaxExpansionNotLess(uint256 newExp, uint256 currentExp);
```
Comment (ceilican) [inline contracts/ContributionAccountingToken.sol:21]: `ExpansionRateNotDecreased` would be a clearer name for this error.

File: contracts/ContributionAccountingToken.sol [hunk 3]
```diff
@@ -69,27 +73,43 @@ contract ContributionAccountingToken is ERC20Burnable, ERC20Permit, AccessContro
     }
 
     function mint(address to, uint256 amount) public onlyRole(MINTER_ROLE) {
-        require(amount <= maxMintableAmount(), "Exceeds maximum mintable amount");
-        
+        uint256 available = maxMintableAmount();
+        if (amount > available) {
```
Comment (ceilican) [inline contracts/ContributionAccountingToken.sol:77]: Keep all these `if ... revert` blocks in a single line.

File: contracts/ContributionAccountingToken.sol [hunk 4]
```diff
@@ -7,11 +7,20 @@ import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit.sol";
...
 
 interface ICATFactory {
-    function grantMinterRole(address catAddress, address minter) external;
     function onMinterRoleGranted(address minter) external;
 }
 
 contract ContributionAccountingToken is ERC20Burnable, ERC20Permit, AccessControl {
+    // Custom errors
+    error ExceedsMaxMintableAmount(uint256 requested, uint256 available);
+    error NewMaxSupplyNotLess(uint256 newMax, uint256 currentMax);
```
[the remark this answers is not in the document]
Comment (DengreSarthak) [reply contracts/ContributionAccountingToken.sol:16]: Done.

File: contracts/ContributionAccountingToken.sol (hunk 2 above)
Comment (DengreSarthak) [reply contracts/ContributionAccountingToken.sol:21]: Done.

**Review by ceilican — APPROVED** [review]
